File: src/main/java/sage/web/HeedController.java
package sage.web;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;
import sage.domain.service.HeedService;
import sage.domain.service.TagService;
import sage.domain.service.UserService;
import sage.transfer.FollowList;
import sage.transfer.FollowListLite;
import sage.transfer.TagCard;
import sage.web.auth.Auth;

import static java.util.stream.Collectors.toList;

@RestController
@RequestMapping(method = RequestMethod.POST)
public class HeedController {
  @Autowired
  private HeedService heedService;
  @Autowired
  private UserService userService;
  @Autowired
  private TagService tagService;
  
  @RequestMapping("/heed/tag/{id}")
  public void heedTag(@PathVariable long tagId) {
    Long uid = Auth.checkCuid();
    heedService.heedTag(uid, tagId);
  }
  
  @RequestMapping("/unheed/tag/{id}")
  public void unheedTag(@PathVariable long tagId) {
    Long uid = Auth.checkCuid();
    heedService.unheedTag(uid, tagId);
  }

  @RequestMapping("/heed/follow-list/{id}")
  public void heedFollowList(@RequestParam long id) {
    Long cuid = Auth.checkCuid();
    heedService.heedFollowList(cuid, id);
  }

  @RequestMapping("/unheed/follow-list/{id}")
  public void unheedFollowList(@RequestParam long id) {
    Long cuid = Auth.checkCuid();
    heedService.unheedFollowList(cuid, id);
  }

  @RequestMapping(value = "/heeds/tag", method = RequestMethod.GET)
  public List<TagCard> tagHeeds() {
    Long cuid = Auth.checkCuid();
    return heedService.tagHeeds(cuid).stream()
        .map(tagHeed -> tagService.getTagCard(tagHeed.getTag().getId())).collect(toList());
  }

  @RequestMapping(value = "/heeds/follow-list", method = RequestMethod.GET)
  public List<FollowList> followListHeeds() {
    Long cuid = Auth.checkCuid();
    return heedService.followListHeeds(cuid).stream()
        .map(flHeed -> FollowListLite.fromEntity(flHeed.getList()).toFull(userService::getUserLabel, tagService::getTagLabel))
        .collect(toList());
  }
}

File: src/main/kotlin/sage/web/api/HeedController.kt
package sage.web.api

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RestController
import sage.service.HeedService
import sage.service.TagService
import sage.transfer.TagCard
import sage.web.auth.Auth

@RestController
class HeedController @Autowired constructor(
  private val heedService: HeedService,
  private val tagService: TagService) {

  @PostMapping("/heed/tag/{id}")
  fun heedTag(@PathVariable id: Long) {
    val uid = Auth.checkUid()
    heedService.heedTag(uid, id)
  }

  @PostMapping("/unheed/tag/{id}")
  fun unheedTag(@PathVariable id: Long) {
    val uid = Auth.checkUid()
    heedService.unheedTag(uid, id)
  }

  @GetMapping("/heeding/tag/{id}")
  fun heedingTag(@PathVariable id: Long) = heedService.existsTagHeed(Auth.checkUid(), id)

  @GetMapping("/heeds/tag")
  fun tagHeeds(): List<TagCard> {
    val cuid = Auth.checkUid()
    return heedService.tagHeeds(cuid).map {
      tagHeed -> tagService.getTagCard(tagHeed.tag.id)
    }
  }
}
